### DISCUSSION TITLE
Быстрая подписка

### DISCUSSION THREAD

**Discussion by Maracumba** [Chimildic/goofy#61] [Личный опыт]
В общем бывает ставишь лойсы и у некоторых артистов накапливается их много, а в подписках их нет, следовательно за новинками не следишь :rolleyes:
Делал такой скрипт для человека, который только пришел в споти с перенесенными лайками, так сказать быстрая подписка. Может кому еще пригодиться, буду рад.

Вписываем сколько лайков должно быть у артиста без подписки в **_const LIKES = ?_**, в итоге подпишемся на всех артистов у кого лайков больше или равно заданому числу, и создадим плейлист чтобы проверить артистов, вдруг они уже не достойны подписки :rofl:
PS и все это было бы невозможно без **создателя Goofy**, спасибо ему огромное!!!

```js
function followArtists() {
    const LIKES = 5;
    const excludeArtists = [
        // { id: '' }, 
        // { id: '' }
    ];

    let loveTracks = Source.getSavedTracks();
    let followArtists = Source.getArtists({ followed_include: true });
    Combiner.push(followArtists, excludeArtists);
    Filter.removeArtists(loveTracks, followArtists);

    let unfollowArtists = findUnfollowArtists();
    Library.followArtists(unfollowArtists);
    createPlaylist();

    function findUnfollowArtists() {
        let unfollowArtists = Object.values(calcRepeat())
            .filter((r) => r.count >= LIKES)
            .map((r) => r.item);
        console.log('Артистов без подписки:', unfollowArtists.length);
        return unfollowArtists;
    }

    function calcRepeat() {
        let loveArtists = loveTracks.map((t) => t.artists[0]);
        return loveArtists.reduce((repeat, artist) => {
            repeat[artist.id] = repeat[artist.id] || { count: 0, item: artist };
            repeat[artist.id].count++;
            return repeat;
        }, {});
    }

    function createPlaylist() {
        let topTracks = Source.getArtistsTopTracks(unfollowArtists, false);
        topTracks.forEach((tracks) => tracks.splice(1, 10));
        topTracks = topTracks.flat(1);
        Playlist.saveWithReplace({
            name: 'Артисты без подписки',
            tracks: topTracks,
            public: false,
        });
    }
}
```

**Comment by amd64fox** [comment]
Спасибо @Chimildic Такой  вопрос назрел, как можно сделать список исключения ?  Например у  группы Ленинград 5 лайков, но я не хочу на неё подписываться.

И ещё,  здесь код вставлен криво.

**Reply by efremych** [reply]
> список исключения

Создать плейлист и добавить в него трек группы Ленинград.
```js
function UnfollowArt() {
const LIKES = 5; //сколько лайков должно быть у артиста
let LoveTracks = Source.getSavedTracks();
let excludeArtists = Source.getPlaylistTracks('Название плейлиста', 'id плейлиста');
Filter.removeArtists(LoveTracks, excludeArtists );
let FollowArt = Source.getArtists({
followed_include: true,
});
Filter.removeArtists(LoveTracks, FollowArt);
```

Или на этапе отбора артистов исключать
https://chimildic.github.io/goofy/#/func?id=getartiststracks

**Reply by Maracumba** [reply]
Спасибо @Chimildic за редактирование кода
Теперь все просто и красиво) Я так не умею, делаю по простому. Хотя тоже охота так же прибрать свои скрипты.

**Reply by amd64fox** [reply]
> Или на этапе отбора артистов исключать

Спасибо, 2-ой вариант, конечно лучше, уже в основном коде @Chimildic  добавил, спасибо еще раз.

**Reply by amd64fox** [reply]
сейчас проверял, все работает кроме исключений почему-то, добавил 3 id артиста в константу `excludeArtists`, но на них скрипт подписывается тоже.

**Reply by Chimildic** [reply]
Потому что этот исполнитель удаляется в `getArtists` и его нет в `followArtists`. Поэтому `removeArtists` не удаляет его из `loveTracks` и он проходит по условию `>= LIKES`. Одно из предложений @efremych выходит ошибочно. Изменил код

**Reply by amd64fox** [reply]
Спасибо, работает.
